Using label index for schema index creation

MishaDemianenko opened the pull request:
Switch to label scan store to determine list of node ids for specified label for index population.
In cases when label is not used to mark significant amount of nodes this gives noticeable performance boost.
For cases when label scan store index does not exist yet - full node scan is used.
Feature is experimental and turn off by default.
To turn it on please see feature flag 'use.label.index' in DynamicIndexStoreView.
To be able to adjust when label either node store should be used please see feature toggle 'all.nodes.visit.percentage.threshold' in DynamicIndexStoreView.

changelog: Add possibility to use label store as a source of affected nodes for schema index creation.

tinwelint committed:
Supports querying multiple label ids AND/OR from LabelScanStoreReader
(cherry picked from commit ff85bc5)

MishaDemianenko committed:
Using labels scan store for index creation instead

Switch to label scan store to determine list of node ids for specified label for index population.
In cases when label is not used to mark significant amount of nodes this gives noticeable performance boost.
For cases when label scan store index does not exist yet - full node scan is used.
Code cleanup.

Handling of concurrent updates druing index population by queueing them till
we finish with nodes available from label scan store.

In case if we use label scan store as source of nodes it's possible to observe following situation:
                           Update
                           for Node8
                             +
+-----+           +-----+    |     +-----+
|Node1+---------> |Node7+--------> |Node9|
+-----+           +-----+          +-----+

In this case update for node 8 will come just after node 7 was processed and in case of sequential node scan should be visible on a next step.
But that's not the case when we use label scan store, since current index reader will not gonna see that update.
To be able to handle those cases and cases when we see some nodes from the part that we already bypass
we will queue all updates till the end of population and will replay them afterwords.

Sample index build using label scan store scan after population completion
using specialized sampler that will perform index commit and will reread all values from index,
since we can't rely on population data.

MishaDemianenko committed:
Fix failing tests

MishaDemianenko commented:
@tinwelint i have update proposal to include updated _idea ready_ code version that should give mostly clean picture of how things working/should work. PR still contains some updates and TODOs that here only temporary/for ongoing testing, but i think if you interesting you can already take a look.

MishaDemianenko committed:
Check for empty label scan store during store view selection.

MishaDemianenko committed:
Restore licenses.

MishaDemianenko committed:
Refactor Adaptable store view - move new methods to store scan. Update tests.

MishaDemianenko committed:
Restore QUEUE_THRESHOLD, javadocs.
Remove unrelated code style changes, renaming stuff.

MishaDemianenko committed:
Introduce feature flag to disable usage of label scan store for index construction.

MishaDemianenko committed:
Update default percentage to be max 10% from number of nodes as allowable threshold for label index usage during index creation.

MishaDemianenko committed:
Cleanup todos and revert import changes

MishaDemianenko committed:
Adapt test expectations to much new threshold percentage.

MishaDemianenko committed:
Update complete phase handling, add more tests;
Adapt tests for new disable by default toggle switch.

burqen commented:
@MishaDemianenko If this change speeds up index creation a lot I think it is suitable for changelog.

MishaDemianenko commented:
@burqen for now idea is that this feature will be disabled by default in old versions

burqen commented:
@MishaDemianenko Guarded by some experimental config option? If we want people to try it out (which is useful for feedback) we should not hide it, instead promote it as an experimental feature to try at own risk.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/state/storeview/LabelScanViewNodeStoreScan.java:
```diff
@@ -0,0 +1,211 @@
+        long nodeId = update.getNodeId();
+        for ( long label : labels )
+        {
+            PrimitiveLongObjectMap<PrimitiveLongSet> labelsNodes = propertyLabelNodes.get( property );
```
This looks like a pretty heavy datastructure, `map<long,set<long>>` could we instead only make due with a `set<long>` of changed nodes and go and read the node labels from the store on completion? That's what e.g. recovery does for reindexing purposes. It would require way less memory.

tinwelint replied:
Another thing which may be good would be to, instead of queueing up _all_ updates during the scan, to look at number of pending updates and if they are rising then the label scan could do this incrementally: i.e. at some point during the scan stop, apply pending updates and create a new reader, resetting the update queue.

MishaDemianenko replied:
second one sounds like a nice suggestion will try that...

tinwelint commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/schema/populator/DirectNonUniqueIndexSampler.java:
```diff
@@ -0,0 +1,99 @@
+    @Override
+    public IndexSample result( int numDocs )
+    {
+        throw new UnsupportedOperationException( "Not implemented." );
```
And there's no point implementing it? Perhaps we could throw something else then... or at least a different message

MishaDemianenko replied:
What other exception you can think of?

tinwelint replied:
Dunno, at least a better message. The current message looks like a TODO

tinwelint commented on enterprise/kernel/src/test/java/org/neo4j/kernel/impl/enterprise/DynamicIndexStoreIT.java:
```diff
@@ -0,0 +1,167 @@
+import org.neo4j.test.SuppressOutput;
+import org.neo4j.test.TargetDirectory;
+
+public class DynamicIndexStoreIT
```
Do this test need to be in enterprise-kernel?

tinwelint replied:
also this test takes 40s on my laptop run, probably much more in CI. We should be wary of our test times. Try to test the gist of what you're testing without exaggerating

MishaDemianenko replied:
that should be somewhere where we actually use lucene to populate indexes, that' why enterprise kernel was chosen. Update test to pass much faster!

tinwelint replied:
Cool, for that we usually put those in community/neo4j

tinwelint reviewed (commented):
Generally looks good.

I don't understand the index sampling changes or problem quite, but I'm OK with that as long as we have tests there.

As a side note one might argue that a more generic approach to deciding whether or not to use label scan store to scan nodes with certain label might in the future be moved to be a statistical thing so that labels which nodes above a certain percentage of all nodes have wouldn't even be indexed in the label scan store and decision to use or not use label scan store could be transparent on a lower level.

MishaDemianenko committed:
Rename and make initial dataset smaller to allow test to pass much faster.

MishaDemianenko committed:
Update label scan store index population to allow updates and perform index reader update each time when observe update that come from non visited yet range of node ids.
Remove introduced complete method from StoreScan interface.

MishaDemianenko committed:
Move test to neo4j module instead of kernel-enterpise

MishaDemianenko committed:
Use batched mode of update population in all multiple index populators.

MishaDemianenko commented:
@tinwelint should be ready for, all comments should be addressed.

MishaDemianenko commented:
@burqen change log and description updated

burqen commented:
@MishaDemianenko Nice! How do I activate feature?

MishaDemianenko commented:
@burqen description mentions that right?

burqen commented:
@MishaDemianenko Sure, but I wont see that in the changelog. Although, there will be a link to the PR in the changelog, so maybe that is good enough. Let's agree it is.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java:
```diff
@@ -131,14 +129,11 @@ protected IndexPopulation createPopulation( IndexPopulator populator,
-            log.debug( "Drained queue and all batched updates." + EOL + this );
-        }
+        log.debug( "Populating from queue." + EOL + this );
+        flushAll();
```
Doing this flushing every time there's an update will produce more and smaller tasks to the executor, won't it? Could it be less efficient to do this? Have you measured?

MishaDemianenko replied:
@tinwelint no no, now flushing condition moved up to MultipleIndexPopulator and we will call `populateFromQueue` now only when `isQueueThresholdReached` will allow us to do that; so in terms of behaviour it should be exactly the same

tinwelint commented:
The simplification where the label scan reader is marked as outdated looks really good to me, there's no refresh involved in closing/re-opening the reader, right?

MishaDemianenko commented:
No, new refreshes where not introduced, we should see updates since they will be applied during commit process and in the end of that index should be refreshed.

MishaDemianenko committed:
Sort label scan store results

Sort results of label scan store index by range to always retrieve nodes in ascending order.
Order of ids is important for example for index population in cases when label scan store is used as matched nodes source.

tinwelint commented:
Neo4j is now relying on results to be sorted from any `LabelScanStore` (`LabelScanReader` that is). I'd like to see that reflected in its javadocs and compatibility tests if such exist

MishaDemianenko commented:
@tinwelint i have added `retrieveNodeIdsInAscendingOrder` that check that we get nodes from label scan store always in ascending order. Or do you have some other compatibility tests in mind?

tinwelint commented:
@MishaDemianenko hmm actually there doesn't seem to be such a test suite, then it's fine. Creating such could be done as separate piece of work. OK I think we're quite done here actually!

MishaDemianenko commented:
LGTM!
